From 546ebeadcf5b9280ffa80b2a9f6bbe52168b786e Mon Sep 17 00:00:00 2001
From: Abhay Singh Gill <abhaygill017@gmail.com>
Date: Sat, 13 Sep 2025 21:32:55 +0530
Subject: [PATCH 4/4] SystemUI: Add flashlight strength control

The default behaviour is toggling the flashlight since we use secondaryClick as default onClick behaviour on small tiles.

Squashed:

    From: NurKeinNeid <nurkeinneid@derpfest.org>
    Date: Fri, 19 Sep 2025 01:42:17 +0200
    Subject: SystemUI: Fix flashlight strength control detection

    Change detection logic from 'mMaxLevel > mDefaultLevel' to 'mMaxLevel > 1'
    to properly detect devices where max and default levels are equal.

    Fixes flashlight strength control not working on some devices.

    Signed-off-by: NurKeinNeid <nurkeinneid@derpfest.org>
    Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>

    From: Abhay Singh Gill <abhaygill017@gmail.com>
    Date: Fri, 19 Sep 2025 18:26:10 +0530
    Subject: SystemUI: Extend flashlight strength logic to flashlight controller

    Previously the flashlight strength was only applied when toggling it from the QS tile, now it works from other SystemUI components as well.

    Signed-off-by: Abhay Singh Gill <abhaygill017@gmail.com>
    Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>

    From: Abhay Singh Gill <abhaygill017@gmail.com>
    Date: Fri, 19 Sep 2025 22:22:13 +0530
    Subject: SystemUI: Allow long click on flashlight tile

    Would be useful for adjusting flashlight strength when the tile is small/ icon only.

    Signed-off-by: Abhay Singh Gill <abhaygill017@gmail.com>
    Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>

    From: ShevT <evgeny.shishkov@gmail.com>
    Date: Tue, 23 Sep 2025 14:10:54 +0300
    Subject: SystemUI: Changing the logic of the flashlight tile

     Single tap - on/off
     Long tap - flashlight strength dialog

    Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>

    From: John Galt <johngaltfirstrun@gmail.com>
    Date: Mon, 29 Dec 2025 14:02:57 -0500
    Subject: SystemUI: make flashlight tile use real levels

    Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>

    From: John Galt <johngaltfirstrun@gmail.com>
    Date: Tue, 30 Dec 2025 09:51:15 -0500
    Subject: SystemUI: make flashlight slider use primitive haptics if supported

    Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>

Change-Id: I8da1d9c80cd2209a60cde524aa58bd251610244b
Signed-off-by: Abhay Singh Gill <abhaygill017@gmail.com>
Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>
---
 .../SystemUI/res/values/custom_strings.xml    |   5 +
 .../FlashlightQuickAffordanceConfig.kt        |   2 +
 .../systemui/qs/tiles/FlashlightTile.java     |  73 +++++++-
 .../tiles/dialog/FlashlightDialogDelegate.kt  | 146 ++++++++++++++++
 .../FlashlightTileDataInteractor.kt           |   1 +
 .../policy/FlashlightController.java          |  10 ++
 .../policy/FlashlightControllerImpl.java      | 159 +++++++++++++++++-
 7 files changed, 391 insertions(+), 5 deletions(-)
 create mode 100644 packages/SystemUI/src/com/android/systemui/qs/tiles/dialog/FlashlightDialogDelegate.kt

diff --git a/packages/SystemUI/res/values/custom_strings.xml b/packages/SystemUI/res/values/custom_strings.xml
index a2abb9b2e1b1..a790c104283b 100644
--- a/packages/SystemUI/res/values/custom_strings.xml
+++ b/packages/SystemUI/res/values/custom_strings.xml
@@ -116,4 +116,9 @@
     <string name="global_action_restart_download_progress">Restarting to download mode\u2026</string>
     <!-- Button to restart SystemUI, within the Restart Options dialog -->
     <string name="global_action_restart_systemui">SystemUI</string>
+
+    <!-- Flashlight Strength -->
+    <string name="flashlight_strength_title">Flashlight Strength</string>
+    <string name="flashlight_strength_turn_off">Turn off</string>
+    <string name="flashlight_strength_turn_on">Turn on</string>
 </resources>
diff --git a/packages/SystemUI/src/com/android/systemui/keyguard/data/quickaffordance/FlashlightQuickAffordanceConfig.kt b/packages/SystemUI/src/com/android/systemui/keyguard/data/quickaffordance/FlashlightQuickAffordanceConfig.kt
index 683c11a88b89..bfcc47c1dbb6 100644
--- a/packages/SystemUI/src/com/android/systemui/keyguard/data/quickaffordance/FlashlightQuickAffordanceConfig.kt
+++ b/packages/SystemUI/src/com/android/systemui/keyguard/data/quickaffordance/FlashlightQuickAffordanceConfig.kt
@@ -116,6 +116,8 @@ constructor(
                             TAG,
                         )
                     }
+
+                    override fun onFlashlightStrengthChanged(level: Int) {}
                 }
 
             flashlightController.addCallback(flashlightCallback)
diff --git a/packages/SystemUI/src/com/android/systemui/qs/tiles/FlashlightTile.java b/packages/SystemUI/src/com/android/systemui/qs/tiles/FlashlightTile.java
index 2b127d60b2be..565e31424d09 100644
--- a/packages/SystemUI/src/com/android/systemui/qs/tiles/FlashlightTile.java
+++ b/packages/SystemUI/src/com/android/systemui/qs/tiles/FlashlightTile.java
@@ -26,8 +26,11 @@ import android.widget.Switch;
 
 import androidx.annotation.Nullable;
 
+import com.android.internal.jank.InteractionJankMonitor;
 import com.android.internal.logging.MetricsLogger;
 import com.android.internal.logging.nano.MetricsProto.MetricsEvent;
+import com.android.systemui.animation.DialogCuj;
+import com.android.systemui.animation.DialogTransitionAnimator;
 import com.android.systemui.animation.Expandable;
 import com.android.systemui.dagger.qualifiers.Background;
 import com.android.systemui.dagger.qualifiers.Main;
@@ -38,11 +41,14 @@ import com.android.systemui.plugins.statusbar.StatusBarStateController;
 import com.android.systemui.qs.QSHost;
 import com.android.systemui.qs.QsEventLogger;
 import com.android.systemui.qs.logging.QSLogger;
+import com.android.systemui.qs.tiles.dialog.FlashlightDialogDelegate;
 import com.android.systemui.qs.tileimpl.QSTileImpl;
 import com.android.systemui.res.R;
+import com.android.systemui.statusbar.phone.SystemUIDialog;
 import com.android.systemui.statusbar.policy.FlashlightController;
 
 import javax.inject.Inject;
+import javax.inject.Provider;
 
 /**
  * Quick settings tile: Control flashlight
@@ -51,8 +57,15 @@ public class FlashlightTile extends QSTileImpl<BooleanState> implements
         FlashlightController.FlashlightListener {
 
     public static final String TILE_SPEC = "flashlight";
+    private static final String FLASHLIGHT_BRIGHTNESS_SETTING = "flashlight_brightness";
+    private static final String INTERACTION_JANK_TAG = "flashlight_strength";
+
     private final FlashlightController mFlashlightController;
 
+    private final Handler mHandler;
+    private final Provider<FlashlightDialogDelegate> mFlashlightDialogProvider;
+    private final DialogTransitionAnimator mDialogTransitionAnimator;
+
     @Inject
     public FlashlightTile(
             QSHost host,
@@ -64,12 +77,17 @@ public class FlashlightTile extends QSTileImpl<BooleanState> implements
             StatusBarStateController statusBarStateController,
             ActivityStarter activityStarter,
             QSLogger qsLogger,
-            FlashlightController flashlightController
+            FlashlightController flashlightController,
+            Provider<FlashlightDialogDelegate> flashlightDialogDelegateProvider,
+            DialogTransitionAnimator dialogTransitionAnimator
     ) {
         super(host, uiEventLogger, backgroundLooper, mainHandler, falsingManager, metricsLogger,
                 statusBarStateController, activityStarter, qsLogger);
+        mHandler = mainHandler;
         mFlashlightController = flashlightController;
         mFlashlightController.observe(getLifecycle(), this);
+        mFlashlightDialogProvider = flashlightDialogDelegateProvider;
+        mDialogTransitionAnimator = dialogTransitionAnimator;
     }
 
     @Override
@@ -80,7 +98,7 @@ public class FlashlightTile extends QSTileImpl<BooleanState> implements
     @Override
     public BooleanState newTileState() {
         BooleanState state = new BooleanState();
-        state.handlesLongClick = false;
+        state.handlesLongClick = true;
         return state;
     }
 
@@ -100,6 +118,11 @@ public class FlashlightTile extends QSTileImpl<BooleanState> implements
 
     @Override
     protected void handleClick(@Nullable Expandable expandable) {
+        handleSecondaryClick(expandable);
+    }
+
+    @Override
+    protected void handleSecondaryClick(@Nullable Expandable expandable) {
         if (ActivityManager.isUserAMonkey()) {
             return;
         }
@@ -115,7 +138,36 @@ public class FlashlightTile extends QSTileImpl<BooleanState> implements
 
     @Override
     protected void handleLongClick(@Nullable Expandable expandable) {
-        handleClick(expandable);
+        if (ActivityManager.isUserAMonkey()) {
+            return;
+        }
+
+        if (mFlashlightController.isStrengthControlSupported()) {
+            Runnable runnable = new Runnable() {
+                @Override
+                public void run() {
+                    SystemUIDialog dialog = mFlashlightDialogProvider.get().createDialog();
+                    if (expandable != null) {
+                        DialogTransitionAnimator.Controller controller =
+                                expandable.dialogTransitionController(
+                                        new DialogCuj(
+                                                InteractionJankMonitor.CUJ_SHADE_DIALOG_OPEN,
+                                                INTERACTION_JANK_TAG));
+                        if (controller != null) {
+                            mDialogTransitionAnimator.show(dialog, controller);
+                        } else {
+                            dialog.show();
+                        }
+                    } else {
+                        dialog.show();
+                    }
+                }
+            };
+
+            mHandler.post(runnable);
+        } else {
+            handleSecondaryClick(expandable);
+        }
     }
 
     @Override
@@ -123,6 +175,7 @@ public class FlashlightTile extends QSTileImpl<BooleanState> implements
         state.label = mHost.getContext().getString(R.string.quick_settings_flashlight_label);
         state.secondaryLabel = "";
         state.stateDescription = "";
+        state.handlesSecondaryClick = mFlashlightController.isStrengthControlSupported();
         if (!mFlashlightController.isAvailable()) {
             state.secondaryLabel = mContext.getString(
                     R.string.quick_settings_flashlight_camera_in_use);
@@ -131,6 +184,15 @@ public class FlashlightTile extends QSTileImpl<BooleanState> implements
             state.icon = maybeLoadResourceIcon(R.drawable.qs_flashlight_icon_off);
             return;
         }
+        if (mFlashlightController.isStrengthControlSupported()) {
+            boolean enabled = mFlashlightController.isEnabled();
+            float percent = mFlashlightController.getCurrentPercent();
+
+            if (enabled) {
+                state.secondaryLabel = Math.round(percent * 100f) + "%";
+                state.stateDescription = state.secondaryLabel;
+            }
+        }
         if (arg instanceof Boolean) {
             boolean value = (Boolean) arg;
             if (value == state.value) {
@@ -166,4 +228,9 @@ public class FlashlightTile extends QSTileImpl<BooleanState> implements
     public void onFlashlightAvailabilityChanged(boolean available) {
         refreshState();
     }
+
+    @Override
+    public void onFlashlightStrengthChanged(int level) {
+        refreshState();
+    }
 }
diff --git a/packages/SystemUI/src/com/android/systemui/qs/tiles/dialog/FlashlightDialogDelegate.kt b/packages/SystemUI/src/com/android/systemui/qs/tiles/dialog/FlashlightDialogDelegate.kt
new file mode 100644
index 000000000000..1a5437381b40
--- /dev/null
+++ b/packages/SystemUI/src/com/android/systemui/qs/tiles/dialog/FlashlightDialogDelegate.kt
@@ -0,0 +1,146 @@
+/*
+ * Copyright (C) 2025 Neoteric OS
+             (C) 2024-2025 The Clover Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.android.systemui.qs.tiles.dialog
+
+import android.content.Context
+import android.os.Bundle
+import android.os.VibrationEffect
+import android.os.Vibrator
+import android.widget.Button
+import android.widget.FrameLayout
+import android.view.ContextThemeWrapper
+import android.view.Gravity
+import kotlin.math.roundToInt
+import androidx.annotation.MainThread
+import com.android.systemui.dagger.qualifiers.Background
+import com.android.systemui.dagger.qualifiers.Main
+import com.android.systemui.res.R
+import com.android.systemui.statusbar.phone.SystemUIDialog
+import com.android.systemui.statusbar.policy.FlashlightController
+import com.android.systemui.util.concurrency.DelayableExecutor
+import com.google.android.material.slider.Slider
+import javax.inject.Inject
+
+class FlashlightDialogDelegate @Inject constructor(
+    private val context: Context,
+    private val systemUIDialogFactory: SystemUIDialog.Factory,
+    @Main private val mainExecutor: DelayableExecutor,
+    @Background private val backgroundExecutor: DelayableExecutor,
+    private val flashlightController: FlashlightController
+) : SystemUIDialog.Delegate {
+
+    private lateinit var slider: Slider
+
+    private val vibrator = context.getSystemService(Vibrator::class.java)
+    private val flashlightMoveHaptic: VibrationEffect =
+        VibrationEffect.get(VibrationEffect.EFFECT_TICK)
+
+    override fun createDialog(): SystemUIDialog = systemUIDialogFactory.create(this)
+
+    override fun beforeCreate(dialog: SystemUIDialog, savedInstanceState: Bundle?) {
+        dialog.setTitle(R.string.flashlight_strength_title)
+        
+        val container = FrameLayout(dialog.context)
+
+        val themedContext = ContextThemeWrapper(dialog.context, 
+                com.google.android.material.R.style.Theme_Material3_DynamicColors_DayNight)
+        slider = Slider(themedContext).apply {
+            layoutParams = FrameLayout.LayoutParams(
+                FrameLayout.LayoutParams.MATCH_PARENT,
+                FrameLayout.LayoutParams.WRAP_CONTENT,
+                Gravity.CENTER
+            )
+        }
+
+        container.addView(slider)
+        dialog.setView(container)
+        dialog.setPositiveButton(R.string.quick_settings_done, null, true)
+        dialog.setNeutralButton(
+            if (flashlightController.isEnabled())
+                R.string.flashlight_strength_turn_off
+            else
+                R.string.flashlight_strength_turn_on,
+            { _, _ ->
+                val newState = !flashlightController.isEnabled()
+                flashlightController.setFlashlight(newState)
+                slider.isEnabled = newState
+                dialog.getButton(SystemUIDialog.BUTTON_NEUTRAL)?.text =
+                    if (newState)
+                        dialog.context.getString(R.string.flashlight_strength_turn_off)
+                    else
+                        dialog.context.getString(R.string.flashlight_strength_turn_on)
+            },
+            false
+        )
+    }
+
+    override fun onCreate(dialog: SystemUIDialog, savedInstanceState: Bundle?) {
+        val maxLevel = flashlightController.getMaxLevel()
+        val numLevels = maxLevel.coerceAtLeast(1)
+        val currentPercent = flashlightController.getCurrentPercent()
+
+        slider.isEnabled = flashlightController.isEnabled()
+        slider.valueFrom = 1f
+        slider.valueTo = numLevels.toFloat()
+        slider.stepSize = 1f
+        slider.value = ((currentPercent * numLevels).roundToInt()).coerceIn(1, numLevels).toFloat()
+        slider.setLabelFormatter { value ->
+            ((value / numLevels) * 100).roundToInt().coerceIn(0, 100).toString() + "%" }
+
+        var lastLevel = -1
+        slider.addOnChangeListener { _, value, fromUser ->
+            if (fromUser) {
+                val level = value.toInt()
+                if (level != lastLevel) {
+                    vibrateForLevel(level, numLevels)
+                    lastLevel = level
+                }
+                updateFlashlightStrength(level, maxLevel)
+            }
+        }
+    }
+
+    private fun vibrateForLevel(level: Int, maxLevel: Int) {
+        val v = vibrator ?: return
+        if (!v.hasVibrator()) return
+
+        val safeMax = maxLevel.coerceAtLeast(1)
+        val safeLevel = level.coerceIn(1, safeMax)
+        val denom = (safeMax - 1).coerceAtLeast(1)
+        val t = ((safeLevel - 1).toFloat() / denom).coerceIn(0f, 1f)
+        val minScale = 0.12f
+        val scale = (minScale + (1.0f - minScale) * t).coerceIn(0f, 1f)
+
+        val primitive = VibrationEffect.Composition.PRIMITIVE_LOW_TICK
+        if (v.areAllPrimitivesSupported(primitive)) {
+            val effect = VibrationEffect.startComposition()
+                .addPrimitive(primitive, scale)
+                .compose()
+            v.vibrate(effect)
+        } else {
+            // Fallback
+            v.vibrate(flashlightMoveHaptic)
+        }
+    }
+
+    @MainThread
+    private fun updateFlashlightStrength(level: Int, maxLevel: Int) {
+        flashlightController.setFlashlightStrengthLevel(
+            level.coerceIn(1, maxLevel.coerceAtLeast(1))
+        )
+    }
+}
diff --git a/packages/SystemUI/src/com/android/systemui/qs/tiles/impl/flashlight/domain/interactor/FlashlightTileDataInteractor.kt b/packages/SystemUI/src/com/android/systemui/qs/tiles/impl/flashlight/domain/interactor/FlashlightTileDataInteractor.kt
index 2fd2486a47c3..ab2bb5baa2b4 100644
--- a/packages/SystemUI/src/com/android/systemui/qs/tiles/impl/flashlight/domain/interactor/FlashlightTileDataInteractor.kt
+++ b/packages/SystemUI/src/com/android/systemui/qs/tiles/impl/flashlight/domain/interactor/FlashlightTileDataInteractor.kt
@@ -70,6 +70,7 @@ constructor(
                                 else FlashlightModel.Unavailable.Temporarily.CameraInUse
                             )
                         }
+			override fun onFlashlightStrengthChanged(level: Int) {}
                     }
                 flashlightController.addCallback(callback)
                 awaitClose { flashlightController.removeCallback(callback) }
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/policy/FlashlightController.java b/packages/SystemUI/src/com/android/systemui/statusbar/policy/FlashlightController.java
index dc1b27cfb1f2..a3746c7cf8f5 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/policy/FlashlightController.java
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/policy/FlashlightController.java
@@ -25,6 +25,13 @@ public interface FlashlightController extends CallbackController<FlashlightListe
     boolean isAvailable();
     boolean isEnabled();
 
+    boolean isStrengthControlSupported();
+    int getMaxLevel();
+    int getDefaultLevel();
+    int getCurrentLevel();
+    float getCurrentPercent();
+    void setFlashlightStrengthLevel(int level);
+
     @WeaklyReferencedCallback
     public interface FlashlightListener {
 
@@ -45,5 +52,8 @@ public interface FlashlightController extends CallbackController<FlashlightListe
          * @param available true if the flashlight is currently available.
          */
         void onFlashlightAvailabilityChanged(boolean available);
+
+        /** Called when the flashlight strength changes (if supported). */
+        void onFlashlightStrengthChanged(int level);
     }
 }
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/policy/FlashlightControllerImpl.java b/packages/SystemUI/src/com/android/systemui/statusbar/policy/FlashlightControllerImpl.java
index 615cc74ef5e4..dda43180f1c9 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/policy/FlashlightControllerImpl.java
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/policy/FlashlightControllerImpl.java
@@ -17,10 +17,13 @@
 package com.android.systemui.statusbar.policy;
 
 import android.annotation.WorkerThread;
+import android.content.Context;
 import android.content.pm.PackageManager;
 import android.hardware.camera2.CameraAccessException;
 import android.hardware.camera2.CameraCharacteristics;
 import android.hardware.camera2.CameraManager;
+import android.os.Handler;
+import android.os.UserHandle;
 import android.provider.Settings;
 import android.provider.Settings.Secure;
 import android.text.TextUtils;
@@ -32,6 +35,7 @@ import com.android.internal.annotations.GuardedBy;
 import com.android.systemui.broadcast.BroadcastSender;
 import com.android.systemui.dagger.SysUISingleton;
 import com.android.systemui.dagger.qualifiers.Background;
+import com.android.systemui.dagger.qualifiers.Main;
 import com.android.systemui.dump.DumpManager;
 import com.android.systemui.util.settings.SecureSettings;
 
@@ -56,11 +60,14 @@ public class FlashlightControllerImpl implements FlashlightController {
     private static final int DISPATCH_ERROR = 0;
     private static final int DISPATCH_CHANGED = 1;
     private static final int DISPATCH_AVAILABILITY_CHANGED = 2;
+    private static final int DISPATCH_STRENGTH_CHANGED = 3;
 
     private static final String ACTION_FLASHLIGHT_CHANGED =
         "com.android.settings.flashlight.action.FLASHLIGHT_CHANGED";
 
+    private final Context mContext;
     private final CameraManager mCameraManager;
+    private final Handler mHandler;
     private final Executor mExecutor;
     private final SecureSettings mSecureSettings;
     private final DumpManager mDumpManager;
@@ -79,16 +86,31 @@ public class FlashlightControllerImpl implements FlashlightController {
     private final AtomicReference<String> mCameraId;
     private final AtomicBoolean mInitted = new AtomicBoolean(false);
 
+    private static final String FLASHLIGHT_BRIGHTNESS_SETTING = "flashlight_brightness";
+
+    @GuardedBy("this")
+    private int mDefaultLevel;
+    @GuardedBy("this")
+    private int mMaxLevel;
+    @GuardedBy("this")
+    private int mCurrentLevel;
+    @GuardedBy("this")
+    private boolean mStrengthControlSupported;
+
     @Inject
     public FlashlightControllerImpl(
+            Context context,
             DumpManager dumpManager,
             CameraManager cameraManager,
+            @Main Handler mainHandler,
             @Background Executor bgExecutor,
             SecureSettings secureSettings,
             BroadcastSender broadcastSender,
             PackageManager packageManager
     ) {
+        mContext = context;
         mCameraManager = cameraManager;
+        mHandler = mainHandler;
         mExecutor = bgExecutor;
         mCameraId = new AtomicReference<>(null);
         mSecureSettings = secureSettings;
@@ -110,7 +132,29 @@ public class FlashlightControllerImpl implements FlashlightController {
     private void tryInitCamera() {
         if (!mHasFlashlight || mCameraId.get() != null) return;
         try {
-            mCameraId.set(getCameraId());
+            String id = getCameraId();
+            if (id != null) {
+                CameraCharacteristics c = mCameraManager.getCameraCharacteristics(id);
+                Boolean flashAvailable = c.get(CameraCharacteristics.FLASH_INFO_AVAILABLE);
+                if (flashAvailable != null && flashAvailable) {
+                    mDefaultLevel = c.get(CameraCharacteristics.FLASH_INFO_STRENGTH_DEFAULT_LEVEL);
+                    mMaxLevel = c.get(CameraCharacteristics.FLASH_INFO_STRENGTH_MAXIMUM_LEVEL);
+                    mStrengthControlSupported = (mMaxLevel > 1);
+                    mCameraId.set(id);
+
+                    if (mStrengthControlSupported) {
+                        float percent = Settings.System.getFloatForUser(
+                                mContext.getContentResolver(),
+                                FLASHLIGHT_BRIGHTNESS_SETTING,
+                                (float) mDefaultLevel / mMaxLevel,
+                                UserHandle.USER_CURRENT);
+                        synchronized (this) {
+                            mCurrentLevel = Math.max(1, Math.round(percent * mMaxLevel));
+                        }
+                        if (DEBUG) Log.d(TAG, "Restored torch level: " + mCurrentLevel);
+                    }
+                }
+            }
         } catch (Throwable e) {
             Log.e(TAG, "Couldn't initialize.", e);
             return;
@@ -131,7 +175,20 @@ public class FlashlightControllerImpl implements FlashlightController {
             synchronized (this) {
                 if (mFlashlightEnabled != enabled) {
                     try {
-                        mCameraManager.setTorchMode(mCameraId.get(), enabled);
+                        if (enabled) {
+                            if (mStrengthControlSupported) {
+                                int level = (mCurrentLevel > 0) ? mCurrentLevel : mDefaultLevel;
+                                level = Math.max(1, Math.min(level, mMaxLevel));
+                                if (DEBUG) Log.d(TAG, "Turning on torch with level " + level);
+                                mCameraManager.turnOnTorchWithStrengthLevel(mCameraId.get(), level);
+                            } else {
+                                if (DEBUG) Log.d(TAG, "Turning on torch (no strength support)");
+                                mCameraManager.setTorchMode(mCameraId.get(), true);
+                            }
+                        } else {
+                            if (DEBUG) Log.d(TAG, "Turning off torch");
+                            mCameraManager.setTorchMode(mCameraId.get(), false);
+                        }
                     } catch (CameraAccessException e) {
                         Log.e(TAG, "Couldn't set torch mode", e);
                         dispatchError();
@@ -153,6 +210,74 @@ public class FlashlightControllerImpl implements FlashlightController {
         return mTorchAvailable;
     }
 
+    public boolean isStrengthControlSupported() {
+        return mStrengthControlSupported;
+    }
+
+    public int getMaxLevel() {
+        return mMaxLevel;
+    }
+
+    public int getDefaultLevel() {
+        return mDefaultLevel;
+    }
+
+    public synchronized int getCurrentLevel() {
+        return mCurrentLevel;
+    }
+
+    public synchronized float getCurrentPercent() {
+        if (!mStrengthControlSupported || mMaxLevel <= 0) {
+            return 0f;
+        }
+        return Math.max(0.01f, (float) mCurrentLevel / (float) mMaxLevel);
+    }
+
+    public void setFlashlightStrengthLevel(int level) {
+        if (!mStrengthControlSupported) {
+            setFlashlight(level > 0);
+            return;
+        }
+        if (mCameraId.get() == null) {
+            mExecutor.execute(this::tryInitCamera);
+        }
+
+        final int requestedLevel = level;
+
+        mHandler.post(() -> {
+            if (mCameraId.get() == null) return;
+
+            int clampedLevel;
+            synchronized (this) {
+                clampedLevel = Math.max(1, Math.min(requestedLevel, mMaxLevel));
+                if (mCurrentLevel == clampedLevel) return;
+                mCurrentLevel = clampedLevel;
+            }
+
+            try {
+                if (DEBUG) Log.d(TAG, "Setting torch strength level: " + clampedLevel);
+                mCameraManager.turnOnTorchWithStrengthLevel(mCameraId.get(), clampedLevel);
+
+                dispatchStrengthChanged(clampedLevel);
+
+                mExecutor.execute(() -> {
+                    float percent = Math.max(0.01f, (float) clampedLevel / mMaxLevel);
+                    Settings.System.putFloatForUser(
+                            mContext.getContentResolver(),
+                            FLASHLIGHT_BRIGHTNESS_SETTING,
+                            percent,
+                            UserHandle.USER_CURRENT
+                    );
+                });
+
+            } catch (CameraAccessException e) {
+                Log.e(TAG, "Couldn't set torch strength level", e);
+                dispatchError();
+            }
+        });
+    }
+
+
     @Override
     public void addCallback(@NonNull FlashlightListener l) {
         synchronized (mListeners) {
@@ -163,6 +288,7 @@ public class FlashlightControllerImpl implements FlashlightController {
             mListeners.add(new WeakReference<>(l));
             l.onFlashlightAvailabilityChanged(isAvailable());
             l.onFlashlightChanged(isEnabled());
+            l.onFlashlightStrengthChanged(getCurrentLevel());
         }
     }
 
@@ -200,6 +326,17 @@ public class FlashlightControllerImpl implements FlashlightController {
         dispatchListeners(DISPATCH_AVAILABILITY_CHANGED, available);
     }
 
+    private void dispatchStrengthChanged(int level) {
+        synchronized (mListeners) {
+            for (WeakReference<FlashlightListener> ref : mListeners) {
+                FlashlightListener l = ref.get();
+                if (l != null) {
+                    l.onFlashlightStrengthChanged(level);
+                }
+            }
+        }
+    }
+
     private void dispatchListeners(int message, boolean argument) {
         synchronized (mListeners) {
             final ArrayList<WeakReference<FlashlightController.FlashlightListener>> copy =
@@ -259,6 +396,24 @@ public class FlashlightControllerImpl implements FlashlightController {
             }
         }
 
+        @Override
+        @WorkerThread
+        public void onTorchStrengthLevelChanged(@NonNull String cameraId, int newStrengthLevel) {
+            if (!TextUtils.equals(cameraId, mCameraId.get())) return;
+            synchronized (FlashlightControllerImpl.this) {
+                if (mCurrentLevel == newStrengthLevel) return;
+                mCurrentLevel = newStrengthLevel;
+            }
+            float percent = Math.max(0.01f, (float) newStrengthLevel / (float) mMaxLevel);
+            Settings.System.putFloatForUser(
+                    mContext.getContentResolver(),
+                    FLASHLIGHT_BRIGHTNESS_SETTING,
+                    percent,
+                    UserHandle.USER_CURRENT);
+            if (DEBUG) Log.d(TAG, "Strength level changed: " + newStrengthLevel + "/" + mMaxLevel);
+            dispatchStrengthChanged(newStrengthLevel);
+        }
+
         private void setCameraAvailable(boolean available) {
             boolean changed;
             synchronized (FlashlightControllerImpl.this) {
-- 
2.34.1

